2018 edition `?` Kleene operator

mark-i-m opened the pull request:
This is my first attempt at implementing the migration lint + 2018 behavior as discussed in #48075

r? @nikomatsakis

mark-i-m commented:
@nikomatsakis Is this close to what you had in mind?

mark-i-m commented:
Also, I think I should add some more tests for 2015 vs 2018 behavior.

alexreg commented:
Nice work, @mark-i-m! Hoping we can get this merged in soon.

mark-i-m commented:
I'm not quite sure I got the lint right... it's currently a straight up warning...

alexreg commented:
@mark-i-m Yes, there look to be 1001 different ways of linting in the compiler code. This is the sort of thing that could *really* do with a chapter in the rustc book!

alexreg commented:
@mark-i-m Oh gosh, I really should have checked before making that comment! The rustc book has come a long way since I last read it. The one thing still really deficient is the macros chapter, sadly... Also, it reminds me I should submit a PR about incremental compilation tests.

nikomatsakis commented on src/libsyntax/ext/tt/quoted.rs:
```diff
@@ -422,10 +422,35 @@ where
                     }
-                    Ok(Ok(op)) => return (Some(token::Question), op),
+                    Ok(Ok(op)) => {
+                        if !features.macro_at_most_once_rep
```
I think the feature gate is not *quite* acting like I expect. My expectation is:

- In Edition 2015:
  - If you do `$()?[*+]`, then you get "`?` as separator".
  - Else, if you do `$(foo)?[^*+]`, treat as "optional":
    - But if feature gate is not enabled, error
- In Edition 2018:
  - If you do `$(foo)?`, then treat as "optional":
    - But if feature gate is not enabled, error.

My goal is that, when it comes time to stabilize, we can just "constant propagate" as if the feature gate was always enabled and get the final behavior we want:

- in Rust 2015 we fallback to old behavior 
- in Rust 2018 we use new behavior all the time

(Is this the plan? Do we have an official, FCP'd proposal somewhere, anyway?)

mark-i-m replied:
> (Is this the plan? Do we have an official, FCP'd proposal somewhere, anyway?)

No, we don't TMK

> in Rust 2015 we fallback to old behavior

Given that this is behavior that will only last for another couple of months, my preference is to
- just keep this unstable on 2015 and only enable in 2018
- make the feature flag always turn on 2018 behavior

mark-i-m replied:
So to clarify:

- In edition 2015:
    - Without feature flag: `?` is a separator. `?` is _not_ a Kleene op.
    - With feature flag: same as 2018
- In edition 2018
    - `?` is _not_ a separator. `?` is a Kleene op that does not accept a separator.

nikomatsakis replied:
I would prefer this:

- In edition 2015:
  - With *or without* feature flag: `?` is a separator and there is no `?` 
  - However, using `?` as a separator issues a `rust_2018_migration` warning
- In edition 2018:
  - `?` is always a separator
    - and feature-gate is required until we stabilize

Point is, as before, that the feature-gate should just be a "error or no error" filter -- the behavior should model what we eventually want when all is said and done otherwise. And we never want Rust 2015 to support `?`.

mark-i-m replied:
> In edition 2018:
>    ? is always a separator 

You mean "in 2018, `?` is always a kleene op", right?

alexreg replied:
(I'd sure hope so heh!)

mark-i-m commented:
@nikomatsakis I think I have implemented what you specified with one major exception: currently, the warning for using `?` as a separator in 2015 edition _always fires_ (i.e. it's not informed by `#[warn()]`).

Do you know how I can hook up the warning to fire when the lint group is set? Also, should it be a `rust_2018_compatibility` warning or a `rust_2018_idioms` warning, as mentioned in https://github.com/rust-lang/rust/pull/50620? My inclination is towards the former.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/395706537) ([raw log](https://api.travis-ci.org/v3/job/395706537/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
    100% |████████████████████████████████| 1.3MB 964kB/s 
Collecting botocore==1.10.45 (from awscli)
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Downloading https://files.pythonhosted.org/packages/2a/c8/b180fb83fa46d2b56ea059177dc3c00647d622987daf5e7ffbc658555ede/botocore-1.10.45-py2.py3-none-any.whl (4.4MB)
    0% |                                | 10kB 43.2MB/s eta 0:00:01
    0% |▏                               | 20kB 18.6MB/s eta 0:00:01
    0% |▎                               | 30kB 23.5MB/s eta 0:00:01
    0% |▎                               | 40kB 14.2MB/s eta 0:00:01
---
[00:23:15]    Compiling syntax_pos v0.0.0 (file:///checkout/src/libsyntax_pos)
[00:23:18]    Compiling rustc_errors v0.0.0 (file:///checkout/src/librustc_errors)
[00:24:14]    Compiling proc_macro v0.0.0 (file:///checkout/src/libproc_macro)
[00:24:25]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[00:24:25] error: `?` is not a macro repetition operator
[00:24:25]    --> librustc/lint/mod.rs:111:52
[00:24:25]     |
[00:24:25] 111 |      $lint_edition: expr => $edition_level: ident $(,)?
[00:24:25] 
[00:24:25] 
[00:24:25] error: expected `*` or `+`
[00:24:25]    --> librustc/lint/mod.rs:111:52
[00:24:25]     |
[00:24:25] 111 |      $lint_edition: expr => $edition_level: ident $(,)?
[00:24:25] 
[00:24:25] 
[00:24:25] error: `?` is not a macro repetition operator
[00:24:25]    --> librustc/lint/mod.rs:125:25
[00:24:25]     |
[00:24:25] 125 |     ($( $lint:expr ),* $(,)?) => {{
[00:24:25] 
[00:24:25] 
[00:24:25] error: expected `*` or `+`
[00:24:25]    --> librustc/lint/mod.rs:125:25
[00:24:25]     |
[00:24:25] 125 |     ($( $lint:expr ),* $(,)?) => {{
[00:24:25] 
[00:24:55] error: aborting due to 4 previous errors
[00:24:55] 
[00:24:55] error: Could not compile `rustc`.
[00:24:55] error: Could not compile `rustc`.
[00:24:55] 
[00:24:55] Caused by:
[00:24:55]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc librustc/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=976f6e3c1eb047f8 -C extra-filename=-976f6e3c1eb047f8 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/release/deps --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-c8176dd12ca6552b.rlib --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-44e761f8b954b499.so --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-22e1fd34ac64b008.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-c8132f366b701b34.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-c8132f366b701b34.rlib --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/s
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

mark-i-m commented:
This still needs some work. I am currently fixing/writing some tests.

mark-i-m commented:
@nikomatsakis I pushed a few tests. Do these tests represent the behavior you were expecting?

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/395969329) ([raw log](https://api.travis-ci.org/v3/job/395969329/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:tidy
tidy check
[00:06:05] * 553 error codes
[00:06:05] * highest error code: E0906
[00:06:05] tidy error: Found 1 features without a gate test.
[00:06:05] Expected a gate test for the feature 'macro_at_most_once_rep'.
[00:06:05] Hint: create a failing test file named 'feature-gate-macro_at_most_once_rep.rs'
[00:06:05]       in the 'ui' test suite, with its failures due to
[00:06:05]       missing usage of #![feature(macro_at_most_once_rep)].
[00:06:05] Hint: If you already have such a test and don't want to rename it,
[00:06:05]       you can also add a // gate-test-macro_at_most_once_rep line to the test file.
[00:06:06] some tidy checks failed
[00:06:06] 
[00:06:06] 
[00:06:06] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:06:06] 
[00:06:06] 
[00:06:06] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:06:06] Build completed unsuccessfully in 0:01:54
[00:06:06] Build completed unsuccessfully in 0:01:54
[00:06:06] Makefile:79: recipe for target 'tidy' failed
[00:06:06] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:02a807e4
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:00f9ac0b:start=1529804763115039036,finish=1529804763120905889,duration=5866853
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:1b0e3cc9
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:09573790
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #51726) made this pull request unmergeable. Please resolve the merge conflicts.

mark-i-m commented:
Ok, I think I've implemented and added tests for everything except that the migration warning is currently an unconditional warning.

All feedback is welcome!

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/396200939) ([raw log](https://api.travis-ci.org/v3/job/396200939/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

mark-i-m commented:
Doh. I forgot that I started using this in places in the compiler :stuck_out_tongue:

mark-i-m commented:
Hmm... some of these are going to be hard to fix...

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/396219215) ([raw log](https://api.travis-ci.org/v3/job/396219215/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:55:52] ---- [run-pass] run-pass/macro-at-most-once-rep.rs stdout ----
[00:55:52] 
[00:55:52] error: compilation failed!
[00:55:52] status: exit code: 101
[00:55:52] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/macro-at-most-once-rep.rs" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/macro-at-most-once-rep/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/macro-at-most-once-rep/auxiliary"
[00:55:52] ------------------------------------------
[00:55:52] 
[00:55:52] ------------------------------------------
[00:55:52] stderr:
[00:55:52] stderr:
[00:55:52] ------------------------------------------
[00:55:52] error: expected `*` or `+`
[00:55:52]    |
[00:55:52]    |
[00:55:52] 24 |     ($($a:ident)? ; $num:expr) => { {
[00:55:52]    |
[00:55:52]    |
[00:55:52]    = note: `?` is not a macro repetition operator
[00:55:52] 
[00:55:52] error: expected `*` or `+`
[00:55:52]    |
[00:55:52] 29 |          )?
[00:55:52]    |           ^
[00:55:52]    |
[00:55:52]    |
[00:55:52]    = note: `?` is not a macro repetition operator
[00:55:52] error: aborting due to 2 previous errors
[00:55:52] 
[00:55:52] 
[00:55:52] ------------------------------------------
---
[00:55:52] 
[00:55:52] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:498:22
[00:55:52] 
[00:55:52] 
[00:55:52] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/File3380992 .
2490436 ./obj/build
1886572 ./obj/build/x86_64-unknown-linux-gnu
729176 ./src
549328 ./obj/build/x86_64-unknown-linux-gnu/stage0
---
143784 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu
143780 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release
138744 ./obj/build/bootstrap/debug/incremental
124172 ./obj/build/bootstrap/debug/incremental/bootstrap-1r3bppl29tbrj
124168 ./obj/build/bootstrap/debug/incremental/bootstrap-1r3bppl29tbrj/s-f2aum0tdwn-13fia35-183djf9a4im2w
107672 ./obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release
103612 ./obj/build/x86_64-unknown-linux-gnu/stage0/lib/ruibcompiler_builtins/modules
32180 ./.git/modules/src/libcompiler_builtins/modules/compiler-rt
31752 ./.git/modules/src/libcompiler_builtins/modules/compiler-rt/objects
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented on src/test/ui/macros/macro-at-most-once-rep-2018-feature-gate.rs:
```diff
@@ -0,0 +1,45 @@
+// Feature gate test for macro_at_most_once_rep under 2018 edition.
+
+// gate-test-macro_at_most_once_rep
+// compile-flags: --edition=2018
```
unrelated, but we should really add a `// edition: 2018` option, shouldn't we?

mark-i-m replied:
Created #51800

mark-i-m replied:
Done!

nikomatsakis commented on src/test/ui/macros/macro-at-most-once-rep-2018-feature-gate.stderr:
```diff
@@ -0,0 +1,71 @@
+error[E0658]: Using the `?` macro Kleene operator for "at most one" repetition is unstable (see issue #48075)
```
Nit: start compiler messages w/ lower-case letters

nikomatsakis commented:
@mark-i-m travis is unhappy, but this code looks good to me overall

mark-i-m commented:
@nikomatsakis Is it ok that the warning is unconditional?

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/396683335) ([raw log](https://api.travis-ci.org/v3/job/396683335/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:48:28] ....................................................................................................
[00:48:33] ....................................................................................................
[00:48:38] ....................................................................................................
[00:48:44] ....................................................................................................
[00:48:50] .......................................................................F............................
[00:49:01] .i..................................................................................................
[00:49:06] ....................................................................................................
[00:49:12] ....................................................................................................
[00:49:19] .......................................i............................................................
[00:49:19] .......................................i............................................................
eature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/ui/macros/macro-at-most-once-rep-2018-feature-gate.rs","byte_start":836,"byte_end":837,"line_start":27,"line_end":27,"column_start":10,"column_end":11,"is_primary":true,"text":[{"text":"    ($(a)?+) => {}","highlight_start":10,"highlight_end":11}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(macro_at_most_once_rep)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: using the `?` macro Kleene operator for \"at most one\" repetition is unstable (see issue #48075)\n  --> /checkout/src/test/ui/macros/macro-at-most-once-rep-2018-feature-gate.rs:27:10\n   |\nLL |     ($(a)?+) => {}\n   |          ^\n   |\n   = help: add #![feature(macro_at_most_once_rep)] to the crate attributes to enable\n\n"}
[00:49:21] {"message":"expected `*` or `+`","code":null,"level":"error","spans":[{"file_name":"/checkout/src/test/ui/macros/macro-at-most-once-rep-2018-feature-gate.rs","byt
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #50966) made this pull request unmergeable. Please resolve the merge conflicts.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/396704627) ([raw log](https://api.travis-ci.org/v3/job/396704627/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:test_run-pass-fulldeps
Check compiletest suite=run-pass-fulldeps mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:01:44] 
[01:01:44] running 94 tests
[01:03:34] ........................................F................................test [run-pass] run-pass-fulldeps/myriad-closures.rs has been running for over 60 seconds
[01:06:47] 
[01:06:47] test result: FAILED. 93 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:06:47] 
[01:06:47] 
[01:06:47] 
[01:06:47] 
[01:06:47] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "3.9.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:06:47] 
[01:06:47] 
[01:06:47] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:06:47] Build completed unsuTue, 26 Jun 2018 05:37:15 GMT
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #51678) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@mark-i-m 

>  Is it ok that the warning is unconditional?

Maybe i'm a bit confused. What do you mean by unconditional? (in both editions? one edition? etc)

mark-i-m commented:
> Maybe i'm a bit confused. What do you mean by unconditional? (in both editions? one edition? etc)

It only happens in 2015 edition, but you cannot turn off the warning.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/397141788) ([raw log](https://api.travis-ci.org/v3/job/397141788/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:test_run-pass-fulldeps
Check compiletest suite=run-pass-fulldeps mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[00:56:32] 
[00:56:32] running 94 tests
[00:58:16] ........................................F....................................test [run-pass] run-pass-fulldeps/myriad-closures.rs has been running for over 60 seconds
$$($$pat:pat)|+)`
[01:00:43]   --> /checkout/src/test/run-pass-fulldeps/auxiliary/procedural_mbe_matching.rs:40:23
[01:00:43]    |
[01:00:43] 40 |     let mbe_matcher = quote_tokens!(cx, $$matched:expr, $$($$pat:pat)|+);
[01:00:43] 
[01:00:43] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[01:00:43] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[01:00:43]   --> /checkout/src/test/run-pass-fulldeps/auxiliary/procedural_mbe_matching.rs:41:23
[01:00:43]    |
[01:00:43] 41 |       let mbe_matcher = quoted::parse(mbe_matcher.into_iter().collect(),
[01:00:43] 42 | |                                     true,
[01:00:43] 42 | |                                     true,
[01:00:43] 43 | |                                     cx.parse_sess,
[01:00:43] 44 | |                                     &Features::new(),
[01:00:43] 45 | |                                     &[]);
[01:00:43] 
[01:00:43] error: aborting due to previous error
[01:00:43] 
[01:00:43] For more information about this error, try `rustc --explain E0061`.
---
[01:00:43] 
[01:00:43] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:498:22
[01:00:43] 
[01:00:43] 
[01:00:43] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "3.9.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:00:43] 
[01:00:43] 
[01:00:43] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:00:43] Build completed unsuccessfully in 0:19:43
[01:00:43] Build completed unsuccessfully in 0:19:43
[01:00:43] Makefile:58: recipe for target 'check' failed
[01:00:43] make: *** [check] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:00ebf578
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented:
> It only happens in 2015 edition, but you cannot turn off the warning.

Oh. We can't make it a lint? I expect this to be a 'migration lint", since it's something that would change behavior in Rust 2018.

mark-i-m commented:
I'm not really sure how to make it a lint since the logic that parses macros-by-example is in `quoted.rs`. It looks kind of hard to write an out-of-line lint that detects the same without just duplicating lots of code.

mark-i-m commented:
Basically what we need is a way to iterate over all kleene ops in all macros to check their separators...

bors commented:
:umbrella: The latest upstream changes (presumably #51859) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/libsyntax/ext/tt/quoted.rs:
```diff
@@ -406,71 +457,147 @@ where
+                    // #2 is a Kleene op, which is the the only valid option
+                    Ok(Ok((op, _))) => {
+                        // Warn that `?` as a separator will be deprecated
+                        sess.span_diagnostic.span_warn(
```
Well, it seems like we could issue a lint here -- we just need an `id` -- presumably we can thread the node-id for the macro definition down here? We would then invoke [`buffer_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html#method.buffer_lint), I believe, since HIR lowering has not yet happened:

```rust
sess.buffer_lint(
    LINT_NAME,
    macro_definition_id,
    op1_span,
    "using `?` as a separator is deprecated",
)
```

then we can put the lint into the migration lint set properly.

mark-i-m replied:
Thanks! I will look into this tomorrow.

mark-i-m replied:
@nikomatsakis I looked into this a bit more. I am running into two problems:
- `buffer_lint` is not defined on `ParseSess`.
- I don't know where/how to `declare_lint!`.  Doing so requires adding a dependency to `libsyntax` on `librustc`. Is that possible? Assuming it is possible, should I declare the lint in rustc::lint::builtins or in syntax::ext::tt::quoted or somewhere else?

nikomatsakis commented:
@mark-i-m to be clear, we never made a formal FCP around this point, did we? I think we should do so. I feel pretty good about using the edition here, though a *bit* nervous that we have no backwards compatibility story if you *were* using `?` as a separator, other than "don't move to the new edition". Maybe the answer is that nobody does that?

mark-i-m commented:
>  to be clear, we never made a formal FCP around this point, did we?

No we did not. Should we propose that now?

>  I feel pretty good about using the edition here, though a bit nervous that we have no backwards compatibility story if you were using ? as a separator, other than "don't move to the new edition".

Yeah, this is the only down-side that I can see with the current proposal. Basically, you need to choose any other separator.

> Maybe the answer is that nobody does that?

I think @oli-obk had mentioned that the original PR (that was rolledback) broke clippy. I expect that there are some uses of `?` as a separator. But not many. Fixing the breakage is pretty easy, though, as you just change the separator. Overall, I think it is worth it to have this feature.

nikomatsakis commented:
@mark-i-m the main problem I see re: breakage is for externally facing macros, but I agree it's worth it overall

nikomatsakis commented:
I think we should create an issue branched off from the tracking issue and do an FCP, yes. I think the idea would be to have a write-up that clarifies:

- What the new behavior is:
  - Edition 2015: `?` is not available, behaves as separator
  - Edition 2018: `?` behaves like `*` would
- Why we chose this behavior (I'm actually not sure of the full arguments, here are some guesses)
  - "Clean choice" -- edition 2018 behaves as you would probably expect
- Alternatives: notably fallback
  - Were there complications here?
  - You mentioned original PR (which I think attempted fallback?) broke something?

Do you want me to open the issue? Do you think you can supply such a write-up? Who are the list of people that should be cc'd so they can add their 2 cents =)

durka commented:
What about a macro exported from a 2015 crate, using `?` as a separator, but used in a 2018 crate? (This is a more general question about macro interoperability, but I don't see the general answer in the edition guide, apologies if it's been answered elsewhere.)

mark-i-m commented:
> Do you want me to open the issue? Do you think you can supply such a write-up? Who are the list of people that should be cc'd so they can add their 2 cents =)

I think I can do this.

Regarding cc-ing people: I suggest cc-ing those on the original RFC thread/tracking issue discussions. Also petrochenkov since they reviewed the first PR.

----

> What about a macro exported from a 2015 crate, using ? as a separator, but used in a 2018 crate? (This is a more general question about macro interoperability, but I don't see the general answer in the edition guide; apologies if it's been answered elsewhere.)

That's a great question. Currently, they would get a compile error. Perhaps that's ok since I imagine the breakage will be relatively rare? Still not great...

Another option I just thought of might be to have some sort of escaping. i.e. you can actually use `?` as a separator (I imagine we should lint heavily against this) in 2018, but you need to escape it somehow...

mark-i-m commented:
Opened the thread here: #51934

bors commented:
:umbrella: The latest upstream changes (presumably #51762) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
Blocked on making a decision in https://github.com/rust-lang/rust/issues/51934

mark-i-m commented:
@nikomatsakis Thanks!

Do you have any ideas about https://github.com/rust-lang/rust/pull/51587#discussion_r199310832?

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/399380649) ([raw log](https://api.travis-ci.org/v3/job/399380649/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:23:23]    Compiling syntax_pos v0.0.0 (file:///checkout/src/libsyntax_pos)
[00:23:26]    Compiling rustc_errors v0.0.0 (file:///checkout/src/librustc_errors)
[00:24:29]    Compiling proc_macro v0.0.0 (file:///checkout/src/libproc_macro)
[00:24:40]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[00:24:41] error: expected `*` or `+`
[00:24:41]   --> librustc/macros.rs:74:85
[00:24:41]    |
[00:24:41] 74 |     (enum $enum_name:path { $( $variant:ident $( ( $($field:ident $(-> $delegate:tt)?),* ) )* ),* $(,)? }) => {
[00:24:41]    |
[00:24:41]    |
[00:24:41]    = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]   --> librustc/macros.rs:74:103
[00:24:41]    |
[00:24:41] 74 |     (enum $enum_name:path { $( $variant:ident $( ( $($field:ident $(-> $delegate:tt)?),* ) )* ),* $(,)? }) => {
[00:24:41]    |
[00:24:41]    |
[00:24:41]    = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]   --> librustc/macros.rs:93:66
[00:24:41]    |
[00:24:41] 93 |     (struct $struct_name:path { $($field:ident $(-> $delegate:tt)?),*  $(,)? }) => {
[00:24:41]    |
[00:24:41]    |
[00:24:41]    = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]   --> librustc/macros.rs:93:76
[00:24:41]    |
[00:24:41] 93 |     (struct $struct_name:path { $($field:ident $(-> $delegate:tt)?),*  $(,)? }) => {
[00:24:41]    |
[00:24:41]    |
[00:24:41]    = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]    --> librustc/macros.rs:107:72
[00:24:41]     |
[00:24:41] 107 |     (tuple_struct $struct_name:path { $($field:ident $(-> $delegate:tt)?),*  $(,)? }) => {
[00:24:41]     |
[00:24:41]     |
[00:24:41]     = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]    --> librustc/macros.rs:107:82
[00:24:41]     |
[00:24:41] 107 |     (tuple_struct $struct_name:path { $($field:ident $(-> $delegate:tt)?),*  $(,)? }) => {
[00:24:41]     |
[00:24:41]     |
[00:24:41]     = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]   --> librustc/macros.rs:86:98
[00:24:41]    |
[00:24:41] 86 |                             $($( __impl_stable_hash_field!($field, __ctx, __hasher $(, $delegate)?) );*)*
[00:24:41]    |
[00:24:41]    |
[00:24:41]    = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]    --> librustc/macros.rs:103:84
[00:24:41]     |
[00:24:41] 103 |                 $( __impl_stable_hash_field!($field, __ctx, __hasher $(, $delegate)?) );*
[00:24:41]     |
[00:24:41]     |
[00:24:41]     = note: `?` is not a macro repetition operator
[00:24:41] 
[00:24:41] error: expected `*` or `+`
[00:24:41]    --> librustc/macros.rs:117:84
[00:24:41]     |
[00:24:41] 117 |                 $( __impl_stable_hash_field!($field, __ctx, __hasher $(, $delegate)?) );*
[00:24:41]     |
[00:24:41]     |
[00:24:41]     = note: `?` is not a macro repetition operator
[00:25:15] error: Could not compile `rustc`.
[00:25:15] 
[00:25:15] Caused by:
[00:25:15] Caused by:
[00:25:15]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc librustc/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=4d03a96572f416f4 -C extra-filename=-4d03a96572f416f4 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/release/deps --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-998809afc03a8106.so --extern backtrace=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libbacktrace-8eafbc58f085b672.rlib --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-1d65b3b5f2314396.rlib --extern byteorder=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libbyteorder-e1dfd51c8b3177b2.rlib --extern chalk_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libchalk_engine-aaa9e2befba6a02a.rlib --extern flate2=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libflate2-c4b4761bf0b6dcbd.rlib --extern fmt_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libfmt_macros-ea6e9b45f399b67b.so --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-272a56f8aad7f7e9.so --extern jobserver=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libjobserver-c8cb2b8c1baf0a6d.rlib --extern lazy_static=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/liblazy_static-f27533b38a757c6d.rlib --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-babaf616bce40eec.rlib --extern parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libparking_lot-ccb3d88d9f914fd9.rlib --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-c2d87a6b8c6962fe.rlib --extern proc_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libproc_macro-7bc63fa4315e1858.so --extern rustc_rayon=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_rayon-348a65d11d8a9aac.rlib --extern rustc_rayon_core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_rayon_core-fa41d79ed310ac32.rlib --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-f063b3fff3eb1317.rlib --
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented:
@mark-i-m 

Ah, I missed these questions before, sorry:

> I looked into this a bit more. I am running into two problems:
> - `buffer_lint` is not defined on `ParseSess`.
> - I don't know where/how to `declare_lint!`.  Doing so requires adding a dependency to > `libsyntax` on `librustc`. Is that possible? Assuming it is possible, should I declare the lint in rustc::lint::builtins or in syntax::ext::tt::quoted or somewhere else?

It seems like the answer to both of these problems might be to extend `ParseSess` with a custom method like `buffer_macro_expansion_lint` that just dumps the relevant span etc. Then we can read the values out of there and invoke `buffer_lint` on them from somewhere in rustc? 

It's definitely not possible to have libsyntax depend on librustc.

mark-i-m commented:
@nikomatsakis Thanks. I will take a look.

Also, is there a way to ask people to stop using this feature in the compiler? Every time I remove them, someone adds more. That's what the travis failures are.

alexreg commented:
@mark-i-m Looks like CI is being really nit-picky, and wants you to change `TODO` to `FIXME`... then you're all done. :-P

alexcrichton commented:
@mark-i-m with https://github.com/rust-lang/rust/issues/51934 looking to be trending towards merging it looks like the last remaining major item on this PR is https://github.com/rust-lang/rust/pull/51587#discussion_r198893102 I think? 

Do you need help implementing that or finishing that? If so, I may be able to help out!

I've tried reviewing this but it looks like there's a lot of rustfmt as well going on which made it difficult for me to find what changed at the beginning :(

bors commented:
:umbrella: The latest upstream changes (presumably #52268) made this pull request unmergeable. Please resolve the merge conflicts.

mark-i-m commented:
> @mark-i-m with #51934 looking to be trending towards merging it looks like the last remaining major item on this PR is #51587 (comment) I think?

Yep, sorry for the delay. I got caught up in a lot of stuff.

> Do you need help implementing that or finishing that? If so, I may be able to help out!

Thanks! So the current plan is to add a `buffered_lints: Lock<Option<LintBuffer>>` field to `ParseSess` (just like `Session`). However, what I am not sure of is how to pass those buffered lints to the `Session` if compilation gets that far. Also, what do I do with buffered lints if the compilation does _not_ get that far? Any hints would help :)

> I've tried reviewing this but it looks like there's a lot of rustfmt as well going on which made it difficult for me to find what changed at the beginning :(

Ugh. Sorry. I keep accidentally doing that. I will undo those changes.

mark-i-m commented:
@alexcrichton I just took another look. It looks like `macro_parser.rs` is the only file that was hit hard by rustfmt. Almost all of those changes are rustfmt. Almost all changes in all the other files are legit. I will undo the changes in macro_parser.rs soon.

mark-i-m commented:
Hmm... it appears that `LintBuffer` is defined in `librustc` and I can't add a dep on that in `libsyntax` :/

Any ideas?

mark-i-m commented:
Ok, so this is my first prototype: https://github.com/rust-lang/rust/compare/master...mark-i-m:parse_sess_buffered_lints

(Still somewhat WIP)

Does this look like the right direction? The biggest problem is that `LintId` is defined in `rustc::lint`, so I cannot `declare_lint!` anywhere in libsyntax and I cannot import declared lints _into_ libsyntax.

One potential solution to that problem would be to define all of the lints via some libsyntax specific type and teach `rustc::lint` a mapping from libsyntax's type.

alexcrichton commented:
@mark-i-m that looks along the right track yeah! One thing I might recommend is perhaps using an enum defined in libsyntax now for the "lint ID" and then in librustc map that to an actual lint id. We can perhaps move lints eventually, but that should solve the problem for now.

Will you have `NodeId` annotations available to you while parsing though? I was under the impression that those were manufactured late after parsing, but I could be wrong! If that's not the case though it should probably be fine to assign the node id to the crate itself. The id here just means how to control the lint levels, and these are crate-wide controls anyway.

bors commented:
:umbrella: The latest upstream changes (presumably #52275) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
This is looking great to me, thanks @mark-i-m!

One thing I couldn't find though was a test that the lint is working, but maybe I missed it?

mark-i-m commented:
Thanks :)

I'm still working on tests... It takes forever to compile on my laptop so I went to bed :P

Will try to take a look later today

mark-i-m commented:
Still working on this. `buffer_lint` is definitely getting called, but for some reason the lints get lost before they can be passed to the `Session`...

Also, there is a test for the lint: `ui/macros/macro-at-most-once-rep-2015-ques-sep.rs`. Currently it is failing because of this issue...

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/403810010) ([raw log](https://api.travis-ci.org/v3/job/403810010/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:49:02] ....................................................................................................
[00:49:06] ....................................................................................................
[00:49:09] ....................................................................................................
[00:49:12] ....................................................................................................
[00:49:15] ..............................................................................................F.....
[00:49:22] ...............................i....................................................................
[00:49:26] ....................................................................................................
[00:49:30] ....................................................................................................
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

mark-i-m commented:
Ok, I think it is ready! :tada:

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/403821623) ([raw log](https://api.travis-ci.org/v3/job/403821623/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:41:17] ....................................................................................................
[00:41:20] ....................................................................................................
[00:41:23] ....................................................................................................
[00:41:26] ....................................................................................................
[00:41:29] ...............................................................................................F....
[00:41:36] ................................i...................................................................
[00:41:40] ....................................................................................................
[00:41:43] ....................................................................................................
[00:41:43] ....................................................................................................
ully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[00:41:49] 
[00:41:49] 
[00:41:49] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[00:41:49] Build completed unsuccessfully in 0:01:21
[00:41:49] Build completed unsuccessfully in 0:01:21
[00:41:49] Makefile:58: recipe for target 'check' failed
[00:41:49] make: *** [check] Error 1
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/403973308) ([raw log](https://api.travis-ci.org/v3/job/403973308/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:test_run-pass-fulldeps
Check compiletest suite=run-pass-fulldeps mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[00:56:26] 
[00:56:26] running 95 tests
[00:58:18] .....................................F............................test [run-pass] run-pass-fulldeps/myriad-closures.rs has been running for over 60 seconds
[00:59:52] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:498:22
[00:59:52] failures:
[00:59:52] 
[00:59:52] ---- [run-pass] run-pass-fulldeps/mbe_matching_test_macro.rs stdout ----
[00:59:52] 
[00:59:52] 
[00:59:52] error: auxiliary build of "/checkout/src/test/run-pass-fulldeps/auxiliary/procedural_mbe_matching.rs" failed to compile: 
[00:59:52] status: exit code: 101
[00:59:52] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/auxiliary/procedural_mbe_matching.rs" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/mbe_matching_test_macro/auxiliary" "-Crpath" "-O" "-Zunstable-options" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/mbe_matching_test_macro/auxiliary"
[00:59:52] ------------------------------------------
[00:59:52] 
[00:59:52] ------------------------------------------
[00:59:52] stderr:
[00:59:52] stderr:
[00:59:52] ------------------------------------------
[00:59:52] warning: unused import: `quote_tokens!(cx, $$matched:expr, $$($$pat:pat)|+)`
[00:59:52]   --> /checkout/src/test/run-pass-fulldeps/auxiliary/procedural_mbe_matching.rs:38:23
[00:59:52]    |
[00:59:52] 38 |     let mbe_matcher = quote_tokens!(cx, $$matched:expr, $$($$pat:pat)|+);
[00:59:52]    |
[00:59:52]    = note: #[warn(unused_imports)] on by default
[00:59:52] 
[00:59:52] error[E0061]: this function takes 7 parameters but 6 parameters were supplied
[00:59:52] error[E0061]: this function takes 7 parameters but 6 parameters were supplied
[00:59:52]   --> /checkout/src/test/run-pass-fulldeps/auxiliary/procedural_mbe_matching.rs:39:23
[00:59:52]    |
[00:59:52] 39 |       let mbe_matcher = quoted::parse(mbe_matcher.into_iter().collect(),
[00:59:52] 40 | |                                     true,
[00:59:52] 40 | |                                     true,
[00:59:52] 41 | |                                     cx.parse_sess,
[00:59:52] 42 | |                                     &Features::new(),
[00:59:52] 43 | |                                     &[],
[00:59:52] 44 | |                                     Edition::Edition2015);
[00:59:52] 
[00:59:52] error: aborting due to previous error
[00:59:52] 
[00:59:52] For more information about this error, try `rustc --explain E0061`.
---
[00:59:52] 
[00:59:52] 
[00:59:52] failures:
[00:59:52]     [run-pass] run-pass-fulldeps/mbe_matching_test_macro.rs
[00:efile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:034c1b42
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
Ok great! I think this is only remaining to be blocked on https://github.com/rust-lang/rust/issues/51934 now.

@mark-i-m one small nit I'd have is that it looks like the lint here is turned on by default in the 2015 edition, but instead could it be tied to the `rust_2018_compatibility` group? That means it would be off-by-default for now, but it would be on-by-default when executing `cargo fix --prepare-for 2018`

mark-i-m commented:
Sure :)

This lint is already part of the group -- it's just that the default needs to be changed.

bors commented:
:umbrella: The latest upstream changes (presumably #52409) made this pull request unmergeable. Please resolve the merge conflicts.

mark-i-m commented:
I'm going to hold off on rebasing again until after FCP...

alexreg commented:
> I'm going to hold off on rebasing again until after FCP...

Just one more day to wait, I think. :-) I'm looking forward to all this work getting merged... it's an awful lot over a small change, but at least everything has been done the "proper" way this time, and it's all backwards compatible, so no one can complain!

mark-i-m commented:
:tada: Rebased :)

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 10ee0f68a6815fafa69f58daf347f0c2a8339f32 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 10ee0f68a6815fafa69f58daf347f0c2a8339f32 -->

bors commented:
:hourglass: Testing commit 10ee0f68a6815fafa69f58daf347f0c2a8339f32 with merge f498e4ec1b57c3245a2fb8e0d5d836ed56760d2d...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.8289), [status-travis](https://travis-ci.org/rust-lang/rust/builds/407657450?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing f498e4ec1b57c3245a2fb8e0d5d836ed56760d2d to master...

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#51587!

Tested on commit rust-lang/rust@f498e4ec1b57c3245a2fb8e0d5d836ed56760d2d.
Direct link to PR: <https://github.com/rust-lang/rust/pull/51587>

💔 rls on windows: test-pass → build-fail (cc @nrc, @rust-lang/infra).
💔 rls on linux: test-pass → build-fail (cc @nrc, @rust-lang/infra).
💔 rustfmt on windows: test-pass → build-fail (cc @nrc, @rust-lang/infra).
💔 rustfmt on linux: test-pass → build-fail (cc @nrc, @rust-lang/infra).

kennytm commented:
```
[01:09:02]   --> /cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-syntax-182.0.0/ext/expand.rs:47:60
[01:09:02]    |
[01:09:02] 47 |             $(one fn $fold_ast:ident; fn $visit_ast:ident;)?
[01:09:02]    |                                                            ^
[01:09:02]    |
[01:09:02]    = note: `?` is not a macro repetition operator
```

We need to update `rustc-ap-syntax`.
